### PR TITLE
Refactor and make the joystick querier re-runnable

### LINKED ISSUE #219: Joystick not getting passed through to games
Hi, I'm trying to use an old Saitek ST290 Pro joystick with some dosbox games.  Dosbox 0.74-3 detects it and passes it through to my games (Testing currently with Wing Commander Privateer), It's detected as a 4axis when joystick type is set to auto.  dosbox-staging claims to see it, but does not pass joystick information on to the game I am testing with.  Thanks for your consideration and time.

Additionally, dosbox-staging does not seem to detect the joystick at all when I set the type manually to 2axis.

### PR DESCRIPTION
**Pull request by krcroft**
**Context:** the joystick was reported unuseable, which was due to the mapper's joystick initialization routine having the following issues:
- It can only be effectively run once because it changes the value of the global variable `joytype` to something other than `AUTO`, which is what this function checks prior to quering SDL's joystick subsystem.     
- When this function is run more than once, it clobbers the prior queried number of joysticks back to zero (without re-requerying for the actual quantities due to locking itself out as mentioned above).  These quantities are used by subsequent code, and when set to zero effectively renders the joystick(s) unuseable.
- Finally, this function depends on several SDL calls to the joystick subsystem, but fails to check if it's initialized.  This is relevant because when the reloadable mapper feature was introduced, it moved forward this joystick initialization routine into the config-file parsing phase, which comes before SDL's initialization in `sdlmain.cpp`;  The calls to SDL's uninitialized joystick routines still passed though - however returning empty values. 

**This PR:**
- Checks for and initializes the joystick subsystem before using it, and properly tears it down on mapper destruction (SDL subsystems are reference-counted)
- Always re-queries the current state of the joysticks, which retains the goal of reloading the mapper during runtime
- Simplifies the joystick initialization logic, but retains the original useability criteria
- Conservatively updates the mapper's joystick global variables only afyer completing the joystick assessment, which protects against early-clobbering if any of the joystick checks fail
- Removes the now redundant Init_Mapper() call from sdlmain, and predicted by @dreamer 
- Adds a small amount of object tracking and cleanup code to stop the mapper from leaking like a sieve (97 leaks fixed)

Thanks to @mdmallardi for catching and reporting this big functional hole, and to @bitstrom for pinpointing the flawed early initialization!

Fixes #219.

### CODE REVIEW

File: src/gui/sdl_mapper.cpp [hunk 1]
```diff
@@ -2362,66 +2367,59 @@ void BIND_MappingEvents(void) {
...
+		SDL_Joystick *stick = SDL_JoystickOpen(i);
+		useable[i] = (SDL_JoystickNumAxes(stick) >= req_min_axis) ||
+		             (SDL_JoystickNumButtons(stick) > 0) ? true : false;
+		SDL_JoystickClose(stick);
+	}
+
+	// Set the type of joystick based which are useable
+	const bool first_usable = useable[0];
+	const bool second_usable = useable[1];	
+	if (first_usable) {
```
Comment (dreamer) [inline src/gui/sdl_mapper.cpp]: Somehow I can't select all the lines 2397-2411, but I am referring to this nested if-structure.

It's already good but could be simplified even a little bit more, to the point where code almost reads like english:
```
if (first_usable && second_usable) {
	joytype = JOY_2AXIS;
	LOG_MSG("SDL: Two or more joysticks found, initializing with 2axis");
} else if (first_usable) {
	joytype = JOY_4AXIS;
	LOG_MSG("SDL: One joystick found, initializing with 4axis");
} else if (second_usable) {
	joytype = JOY_4AXIS_2;
	LOG_MSG("SDL: One joystick found, initializing with 4axis_2");
} else {
	joytype = JOY_NONE;
	LOG_MSG("SDL: No joysticks found");
}
```

Comment (krcroft) [reply src/gui/sdl_mapper.cpp]: This is oh so sweet; thanks for teasing it apart @dreamer!

**Comment by krcroft** [comment]
Added a couple more tiny non-functional commits to mop up the remaining issues reported by PVS Studio.

**Comment by dreamer** [comment]
I just tested changing keymapper file in runtime, and - unfortunately - this change does break this feature :(

My testcase:
- Ctrl-f1, save (to create default mapper file)
- copy this new mapper file to a new location
- Ctrl-f1, select "fullscreen", change the binding to "ctrl-alt-enter" (mod1, mod2), save
- Use following command to reload keymap within dosbox:

      Z:\> config -set "sdl mapperfile=~/.dosbox/copied.map"

- After that, `config -get` will say that new file is being used, but the keys mappings are not reloaded; can be tested if ctrl-alt-enter or alt-enter switches to fullscreen.

**Comment by krcroft** [comment]
Those were perfect instructions @dreamer - thanks! 
Before simply starting DOSBox to trigger the "Reload()" function, I wasn't aware of how to re-run this at runtime.  But using your notes, I see the reload occurs on the `config -set` command.  Was able to reproduce it, and pushed and update. The behavior matches `master` now.

**Comment by krcroft** [comment]
Wrestling w/ a couple static analysis fixes (hopefully this current push satisfies Clang).

It's still working great w/ my game controller here, and passes the reload test per your notes - so I think we're clear to merge. 

Happy if @mdmallardi wants to do more testing after it's merged; but for now, I think it satisfies the issue.

**Comment by dreamer** [comment]
Let's merge this in then :) Thanks!

**Pull request merged**